enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,877 @@
+Only 64 items in request and response header list will be allowed.
+Minimum length of the header value would have to be greater than 0 and less than 4096.
+Sample fetchers for response dynamic header value allowed are "res.hdr" and "ssl_c_der".
+Sample fetchers for request dynamic header value allowed are "req.hdr" and "ssl_c_der".
+Converters allowed for request and response are "lower" and "base64".
+Actions like Set/Delete for headers can't be set for passthrough routes.
+
+D) route/v1 of openshift/api will contain the validations using kubebuilder tags and CEL.
+The validations will be same as mentioned in point C).
+
+E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
+The validations will be same as mentioned in point C) except the one for passthrough routes.
+
+F) To validate the header value it has to match a regex.
+
+For ingress operator CR and route CR(microshift) this is defined on the request and response list of header in openshift/api.
+For route CR(OCP) this is defined in library-go.
+
+This will be the regex:
+```sh
+request header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
+```sh
+response header:
+`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
+```
+
+Allowed dynamic values for request would be -
+```conf
+%[req.hdr(host),lower]
+%[req.hdr(host),base64]
+%[req.hdr(Host)] if foo
+```
+Allowed dynamic values for response would be -
+```conf
+%[res.hdr(X-Acl-Key),lower]
+%[res.hdr(X-Acl-Key),base64]
+%[req.hdr(Host)] if foo
+```
+Dynamic values allowed for both request and response would be -
+```conf
+%{+Q}[ssl_c_der,base64]
+```
+
+For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`
+present in `pkg/route/validation/validation_test.go`
+
+Please note that all values will be sanitized before passing them to the haproxy.conf where a single quote present in value is escaped and the whole value is single quoted which will help
+haproxy not to crash in the event of an invalid input. This sanitize logic will be present in router.
+Quote using single quotes, we will do the following:
+Replace ' with '\''.
+Wrap the entire string in single quotes.
+eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''
+
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+- Would this enhancement cover setting or deleting request headers ?
+  Answer: Yes we would be covering setting or deleting of request headers.
+
+- Would we be able to set load balancing algorithm based on headers ?
+  Answer: No, this out of scope for this enhancement so, it will be considered as a non-goal. Please check point number 5 in the non-goals section for more details.
```

Comment:
It looks like the numbers changed, and you mean to point to what is now number 2 ("Setting load balancing algorithm using header values.").  It seems odd to me that you put the detailed information in the non-goals section and then added a reference here, but the explanation in the non-goals section is clear enough.  
```suggestion
- Will this enhancement cover setting the load-balancing algorithm based on headers?
  Answer: No, this out of scope for this enhancement, so it will be considered as a non-goal. Please check point number 2 in the non-goals section for more details.
```

Reply:
done